schema.go

```diff
@@ -0,0 +1,149 @@
+package tf
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+const (
+	NestingModeSingle NestingMode = 0
+	NestingModeList   NestingMode = 1
+	NestingModeSet    NestingMode = 2
+	NestingModeMap    NestingMode = 3
+)
+
+// Schema is used to define the shape of practitioner-provider information,
+// like resources, data sources, and providers. Think of it as a type
+// definition, but for Terraform.
+type Schema struct {
+	// Attributes are the fields inside the resource, provider, or data
+	// source that the schema is defining. The map key should be the name
+	// of the attribute, and the body defines how it behaves. Names must
+	// only contain lowercase letters, numbers, and underscores.
+	Attributes map[string]Attribute
+
+	// Version indicates the current version of the schema. Schemas are
+	// versioned to help with automatic upgrade process. Whenever you have
+	// a change in the schema you'd like to provide a manual migration for,
+	// you should increment that schema's version by one.
+	Version int64
+}
+
+// Attribute defines the constraints and behaviors of a single field in a
+// schema. Attributes are the fields that show up in Terraform state files and
+// can be used in configuration files.
+type Attribute struct {
+	// Type indicates what kind of attribute this is. You'll most likely
+	// want to use one of the types in the types package.
+	//
+	// If Type is set, Attributes cannot be.
+	Type AttributeType
+
+	// Attributes can have their own, nested attributes. This nested map of
+	// attributes behaves exactly like the map of attributes on the Schema
+	// type.
+	//
+	// If Attributes is set, Type cannot be.
+	//
+	// TODO: do we need MaxItems/MinItems? Can we just make those weird
+	// validation helpers?
+	Attributes map[string]Attribute
+
+	// AttributesNestingMode controls the various ways these sub-groups of
+	// attributes can behave. It can only be used with Attributes, and must
+	// not be set if Type is set.
+	AttributesNestingMode NestingMode
```

Comment:
My initial feeling when seeing these three conflicting options was that this feels a little confusing for provider developers and hard to setup for static analysis. Is there a reason to shy away from always requiring `Type`  (embedding the nesting mode if necessary into those implementations) and treating `Attributes` like the old `Elem`?

Reply:
I originally wanted to combine everything into one attribute, but I eventually gave up on it when I realized I was leading provider developers into confusion.

For example, I considered making `Attributes` fill the `Type` interface.  But that would mean that provider developers could try and set Optional/Computed/Sensitive on elements in a list, which isn't supported. This ends places like hashicorp/terraform-plugin-sdk#201, and I think one of the necessary complexities I'm interested in exposing provider developers to with the framework is the difference between a nested attribute and a complex type, and how those decisions have consequences on what capabilities are available to them.

That's why I thought `Type` and `Attributes` should be separate and mutually exclusive--you need to opt into either this is a cty value, where the hole thing shares one set of flags, or this is a set of nested attributes, where each can have its own individual flags. I think there are benefits to each of those, and they're appropriate in different situations, and so I think the provider developer has to decide which set of trade-offs makes sense for their specific situation.

(`AttributeNestingMode` we can definitely get rid of, I just gotta think my way through how to do that properly.)

Reply:
I think bflad's suggestion might be the same as mine [above](https://github.com/hashicorp/terraform-plugin-framework/pull/11#issuecomment-834296696), which is not that `Attributes` and `Type` should be combined, but that `Type` and `NestingMode` should be combined. Then `Type` would always be required, and some values of `Type` (like `NestedBlock`, say), would also require `Attributes` to be set.

Reply:
Sorry, I didn't read the whole comment and just saw you acknowledged in the last paragraph that we might be able to get rid of nesting modes.